File: LeetCodeSolutions/LeetCode_0739.py
from typing import List


class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        """
        Use stack
        """
        ans = [0] * len(T)
        stack = []
        for _ in range(len(T) - 1, -1, -1):
            while stack and T[_] >= T[stack[-1]]:
                stack.pop()
            if stack:
                ans[_] = stack[-1] - _
            stack.append(_)
        return ans

File: LeetCodeSolutions/LeetCode_0739.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        int n = T.size();
        vector<int > res(n, 0);
        vector<pair<int, int>> stack;
        for(int i = n - 1; i > -1; --i){
            while (!stack.empty() && T[i] >= stack.back().first) stack.pop_back();
            if(stack.empty()) res[i] = 0;
            else res[i] = stack.back().second - i;
            stack.push_back(make_pair(T[i], i));
        }
        return res;

    }
};